### PR TITLE
feat: TileStack - A collection of tiles from a `RenderableTiledMap` for animating

### PR DESCRIPTION
**Pull request by jtmcdole**
# Description

**NOTE: For consideration / conversation going on in Discord.**

Are you building an RPG with a Tiled map - orthographic, isometric, or hexagonal - and want to add extra immersion to your story telling... perhaps a train is rumbling through the map, or your characters have a ground-pound attack and you want the affected tiles to shake with fury... `TileStack` helps you do that in a very flame fashion. 

Example video showing flame-engine animations on all tiles or specific tiles in a layer:

[tiledstack_jiggle_shake.webm](https://user-images.githubusercontent.com/1924313/190840511-1d694022-c6fb-477f-8b6b-56edd119153c.webm)

As simple as:

```dart
    final stack = map.tileMap.tileStack(0, 0, all: true);
    stack.add(
      SequenceEffect(
        [
          MoveEffect.by(
            Vector2(5, 0),
            NoiseEffectController(duration: 1, frequency: 20),
          ),
          MoveEffect.by(Vector2.zero(), LinearEffectController(2)),
          MoveEffect.by(
            Vector2(0, 10),
            NoiseEffectController(duration: 1, frequency: 10),
          ),
        ],
        infinite: true,
      ),
    );
    add(stack);

    final stack2 = map.tileMap.tileStack(4, 0, named: {'floor_under'});
    stack2.add(
      SequenceEffect(
        [
          MoveEffect.by(
            Vector2(5, 0),
            NoiseEffectController(duration: 1, frequency: 20),
          ),
          MoveEffect.by(Vector2.zero(), LinearEffectController(2)),
          MoveEffect.by(
            Vector2(0, 10),
            NoiseEffectController(duration: 1, frequency: 10),
          ),
        ],
        repeatCount: 3,
      )..onComplete = () => stack2.removeFromParent(),
    );
    add(stack2);
```

What is changed:

1. `dart:ui` `RSTransform` is immutable... and SpriteBatch uses it... I need get around that, so I crafted an implementation for Tiled.
2. To get around modifying SpriteBatch, I cache the new `RSTransforms` in the render layers. These are generated lists and only built once - unless you are changing tile data (that's a different problem that should be updated.
3. `TileStack` only implements `PositionProvider`, so it can only get move effects right now.

Notes:
1. Removing the "stack" from your game **does not remove the tiles from the map** - `TileStack` is just a proxy.  It _could_ cause the effect to leave the tiles in a dirty state... so don't do that.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples`.

## Breaking Change

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flame_tiled/lib/src/renderable_tile_map.dart [hunk 1]
```diff
@@ -107,6 +110,74 @@ class RenderableTiledMap {
...
+        x,
+        y,
+        named: named,
+        ids: ids,
+        all: all,
+      ),
+    );
+  }
+
+  /// Recusrive support for [tileStack]
```
Comment (spydon) [inline packages/flame_tiled/lib/src/renderable_tile_map.dart:140]:
```suggestion
  /// Recursive support for [tileStack].
```

File: packages/flame_tiled/lib/src/renderable_tile_map.dart [hunk 2]
```diff
@@ -896,3 +984,86 @@ class _Transform {
...
+
+/// A mutable version of [ui.RSTransform] for custom batch manipulation.
+class _RSTransform implements ui.RSTransform, PositionProvider {
+  final _values = Float32List(4);
+
+  /// This is a cache of `-scos * anchorX + ssin * anchorY`
+  final double _anchorX;
+
+  /// This is a cache of `-ssin * anchorX - scos * anchorY`
+  final double _anchorY;
```
Comment (spydon) [inline packages/flame_tiled/lib/src/renderable_tile_map.dart:996]: Couldn't these quite easily be replaced by `AnchorProvider`?

**Review by spydon — COMMENTED** [review]
Very cool!

**Review by spydon — APPROVED** [review]

**Comment by spydon** [comment]
Oops didn't mean to approve it since it isn't fully done yet, but so far so good!

**Comment by jtmcdole** [comment]
> Oops didn't mean to approve it since it isn't fully done yet, but so far so good!

Any additions, other than tests, you would want?

File: packages/flame_tiled/lib/src/renderable_tile_map.dart (hunk 1 above)
Comment (jtmcdole) [reply packages/flame_tiled/lib/src/renderable_tile_map.dart:140]: ... I need to turn my spellchecker back on.

**Comment by spydon** [comment]
> Any additions, other than tests, you would want?

Docs for the docs site would be awesome, we should get better with that for the bridge libraries (there is a little bit for flame_tiled already I think).

File: packages/flame_tiled/lib/src/renderable_tile_map.dart (hunk 2 above)
Comment (jtmcdole) [reply packages/flame_tiled/lib/src/renderable_tile_map.dart:996]: TL/DR: I don't think we should.

`Anchor` is relative (0.0 to 1.0) - whereas `RSTransform` anchor is a coordinate in the Atlas image (could be a single source image, could be a giant atlas - which now that I look at Orthogonal, it's probably wrong for rotation!). 

`Anchor`:
```dart
  /// The relative x position with respect to the object's width;
  /// 0 means totally to the left (beginning) and 1 means totally to the
  /// right (end).
```

`RSTransform` calculations that are added:

```
The x coordinate of the translation, minus [scos](https://api.flutter.dev/flutter/dart-ui/RSTransform/scos.html) multiplied by the x-coordinate of the rotation point, plus [ssin](https://api.flutter.dev/flutter/dart-ui/RSTransform/ssin.html) multiplied by the y-coordinate of the rotation point.
```


Right now the anchor is hardcoded to the source middle: 

```dart
        final anchorX = src.width / 2; // I think this is wrong for "atlas" with more than 1 image.
// or
        final anchorX = src.width - halfMapTile.x; // I think this is correct.
```

Lastly: what would they ultimately animate and modify? Let's assume we cache the src rectangle and convert anchor to relative - then yes, you could add an [AnchorEffect](https://docs.flame-engine.org/1.3.0/flame/effects.html#anchorbyeffect), but I guess I'm not seeing the usefulness of that with Tiles.

I am happy to be wrong. If you want me to dig into it, I think we could do that as another PR to add more functionality. Thoughts?

**Comment by jtmcdole** [comment]
> > Any additions, other than tests, you would want?
> 
> Docs for the docs site would be awesome, we should get better with that for the bridge libraries (there is a little bit for flame_tiled already I think).

Yes.  I was kinda hoping for a doc's push before next flame_tile release to talk about all the new features we're trying to add.

**Comment by jtmcdole** [comment]
> > Any additions, other than tests, you would want?
> 
> Docs for the docs site would be awesome, we should get better with that for the bridge libraries (there is a little bit for flame_tiled already I think).

docs added in latest commit

File: doc/flame/components.md [hunk 3]
```diff
@@ -911,13 +911,41 @@ void main() {
 
 ## TiledComponent
```
Comment (spydon) [inline doc/flame/components.md:912]: In the long run this section should be moved to `/doc/flame_tiled/`, just like with `flame_audio` for example.
But this doesn't have to be done in this PR if you don't want to.

File: doc/flame/components.md [hunk 4]
```diff
@@ -911,13 +911,41 @@ void main() {
...
+Supported map types include: Orthogonal, Isometric, Hexagonal, and Staggered.
+
+Orthogonal | Hexagonal             |  Isomorphic
+:--:|:-------------------------:|:-------------------------:
+![An example of an orthogonal map](../images/orthogonal.png)|![An example of hexagonal map](../images/pointy_hex_even.png) |  ![An example of isomorphic map](../images/tile_stack_single_move.png)
 
 An example of how to use the API can be found
 [here](https://github.com/flame-engine/flame_tiled/tree/main/example).
 
+Once a `TiledComponent` is loaded, you can select any column of (x,y) tiles in a `tileStack` to
```
Comment (spydon) [inline doc/flame/components.md:928]: I think that this could be it's own subsection

Comment (jtmcdole) [reply doc/flame/components.md:928]: Should flame_tiles have its own doc breakout?

Comment (jtmcdole) [reply doc/flame/components.md:930]: Created a sub-section for TileStack. Agreed we should have a whole doc section for flame_tiled.

**Comment by jtmcdole** [comment]
rebased on main

**Comment by jtmcdole** [comment]
rebased again :D

**Comment by jtmcdole** [comment]
Merge completed.

**Review by spydon — APPROVED** [review]

**Pull request merged**